eclair-core/src/main/scala/fr/acinq/eclair/router/Validation.scala

```diff
@@ -258,29 +257,36 @@ object Validation {
     }
   }
 
-  def handleChannelUpdate(d: Data, db: NetworkDb, routerConf: RouterConf, origins: Set[GossipOrigin], u: ChannelUpdate, wasStashed: Boolean = false)(implicit ctx: ActorContext, log: LoggingAdapter): Data = {
+  def handleChannelUpdate(d: Data, db: NetworkDb, routerConf: RouterConf, update: Either[LocalChannelUpdate, RemoteChannelUpdate], wasStashed: Boolean = false)(implicit ctx: ActorContext, log: LoggingAdapter): Data = {
     implicit val sender: ActorRef = ctx.self // necessary to preserve origin when sending messages to other actors
-    val remoteOrigins = origins flatMap {
-      case r: RemoteGossip if wasStashed =>
-        Some(r.peerConnection)
-      case RemoteGossip(peerConnection, _) =>
-        peerConnection ! TransportHandler.ReadAck(u)
-        log.debug("received channel update for shortChannelId={}", u.shortChannelId)
-        Some(peerConnection)
-      case LocalGossip =>
-        log.debug("received channel update from {}", ctx.sender)
-        None
+    val (u, origins, remoteOrigins) = update match {
+      case Left(lcu) => (lcu.channelUpdate, Set[GossipOrigin](LocalGossip), Nil)
+      case Right(rcu) =>
+        val remote = rcu.origins.collect {
+          case RemoteGossip(peerConnection, _) if wasStashed => peerConnection
+          case RemoteGossip(peerConnection, _) =>
+            peerConnection ! TransportHandler.ReadAck(rcu.channelUpdate)
+            log.debug("received channel update for shortChannelId={}", rcu.channelUpdate.shortChannelId)
+            peerConnection
+        }.toSeq
+        (rcu.channelUpdate, rcu.origins, remote)
     }
     if (d.channels.contains(u.shortChannelId)) {
       // related channel is already known (note: this means no related channel_update is in the stash)
       val publicChannel = true
       val pc = d.channels(u.shortChannelId)
+      val pc1 = update match {
+        case Left(lcu) => pc.updateChannelUpdateSameSideAs(u).updateBalances(lcu.commitments)
+        case Right(_) => pc.updateChannelUpdateSameSideAs(u)
+      }
```

Comment:
We can't do that, we need to use the "old" pc in most places.
I put the computation of `pc1` here to avoid duplication, but I can move it to a helper function which will probably be clearer.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1382/commits/b940a675c019abe022d28572bfd3e0b61dc7cf70, I think it's clearer